oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/api/binary/BinaryHttpUpload.java

```diff
@@ -0,0 +1,163 @@
+ *             <li>larger than fileSize divided by partURLCount</li>
+ *         </ol>
+ *     </li>
+ *     <li>otherwise all part URLs are to be used and the partSize = fileSize divided by partURLCount
+ *     (integer division, discard modulo which will be the last part)</li>
+ *     <li>upload: segment the binary into partSize, for each segment take the next URL from partURLs
+ *     (strictly in order!), proceed with a standard HTTP PUT for each, and for the last part use whatever
+ *     segment size is left</li>
+ *     <li>if a segment fails during upload, retry (up to a certain time out)</li>
+ *     <li>after the upload has finished successfully, notify the application, for example through a
+ *     complete request, passing the {@link #getUploadToken() upload token}, and the application will have to
+ *     call {@link HttpBinaryProvider#completeHttpUpload(String)} with the token</li>
+ * </ol>
+ *
+ * <h2>JSON view</h2>
+ *
+ * A JSON representation of this interface as passed back to a remote client might look like this:
+ * <pre>
+ * {
+ *     "uploadToken": "aaaa-bbbb-cccc-dddd-eeee-ffff-gggg-hhhh",
+ *     "minPartSize": 10485760,
+ *     "maxPartSize": 104857600,
+ *     "partURLs": [
+ *         "http://server.com/upload/1",
+ *         "http://server.com/upload/2",
+ *         "http://server.com/upload/3",
+ *         "http://server.com/upload/4"
+ *     ]
+ * }
+ * </pre>
+ * </p>
+ *
+ */
+public interface BinaryHttpUpload {
+
+    /**
+     * Returns a token identifying this upload. This is required to finalize the upload
+     * at the end by calling {@link HttpBinaryProvider#completeHttpUpload(String)}.
+     *
+     * <p>
+     * The format of this string is depending on the implementation and opaque to consumers.
+     * Implementations must ensure that clients cannot spoof tokens.
+     *
+     * @return a string token identifying this particular upload
+     */
+    String getUploadToken();
+
+    /**
+     * Return the smallest possible part size in bytes. If a consumer wants to choose a custom
+     * part size, it cannot be smaller than this value.
+     *
+     *
+     * @see #getMaxPartSize() getMaxPartSize() for the maximum
+     *
+     * @return smallest possible part size in bytes
+     */
+    long getMinPartSize();
+
+    /**
+     * Return the largest possible part size in bytes. If a consumer wants to choose a custom
+     * part size, it cannot be larger than this value. If this returns -1, the maximum is
+     * unlimited.
+     *
+     * @see #getMinPartSize() getMinPartSize() for the minimum
+     *
+     * @return largest possible part size in bytes or -1 if unlimited
+     */
+    long getMaxPartSize();
+
+    /**
+     * Returns one or more URLs for uploading the binary in one or more parts.
+     *
+     * <p>
+     * In case of multiple URLs, the order in which they are returned is important and defines
+     * the order of the parts. Each part covers a given range of the binary in bytes.
```

Comment:
We should document the requirements for the requests to make using these URLs.

* standard HTTP `PUT`
* must set correct `Content-Length`, `Date` and `Host` header
* body includes the range of bytes for the corresponding part (or the entire binary if a single part)
* can set `Content-Type` if type is known (though only for single PUT uploads)

Reply:
Added - please check to see if it meets your expectations.

Reply:
Thanks.

1. I think exported APIs in oak-jcr (`HttpBinaryProvider`) should not depend on oak-api and it should have its own version of these exceptions. If we want to move that API to jackrabbit-api we _must_ do that since it has no dependency on oak-api.

2. There should be some javadoc on the exceptions.